pubsub/pstest/fake.go

```diff
@@ -954,10 +979,15 @@ func (s *subscription) pull(max int) []*pb.ReceivedMessage {
 	now := s.timeNowFunc()
 	s.maintainMessages(now)
 	var msgs []*pb.ReceivedMessage
-	for _, m := range s.msgs {
+	for id, m := range s.msgs {
 		if m.outstanding() {
 			continue
 		}
+		if s.deadletterCandidate(m) {
+			s.ack(id)
+			s.publishToDeadletter(m)
+			continue
+		}
```

Comment:
Sync delivery attempt field here before incrementing if dead letter policy is set.

```suggestion
		}
		if s.proto.DeadLetterPolicy != nil {
			m.proto.DeliveryAttempt = int32(*m.deliveries)
		}
```